### PR TITLE
Allow the same label to be used in two different units within the same family but in different systems

### PR DESCRIPTION
**Pull request by bsteinbk**
- Allow the same label to be used in two different units within the same family but in different systems.
- This allows 'ft' to be used when parsing Units.FT and Units.SURVEY_FT. Requested by Civil team.
- Add unit test.
- Updated the method Parser.parseIntoQuantity, not currently used in Frontend or UI code, to fix a bug where parsing was not based on the units defined in the Format object.
- Updated UnitsProvide to add ability to find a Unit by label that matches a specific unit system.

### CODE REVIEW

**Review by pmconne — APPROVED** [review]

File: core/quantity/src/Interfaces.ts [hunk 1]
```diff
@@ -38,6 +38,8 @@ export interface UnitConversionSpec {
   name: string;
   /** The default label that is used to display unit */
   label: string;
+  /** Unit system name, used to when finding preferred parse unit */
+  system: string;
```
Comment (calebmshafer) [inline core/quantity/src/Interfaces.ts:42]: We should probably add a note in the NextVersion.md about this as it is a breaking change even though it's a `alpha` API.

File: core/frontend/src/BasicUnitsProvider.ts [hunk 2]
```diff
@@ -17,19 +17,28 @@ import {
...
-  public async findUnit(unitLabel: string, unitFamily?: string): Promise<UnitProps> {
+  public async findUnit(unitLabel: string, unitFamily?: string, unitSystem?: string): Promise<UnitProps> {
+    const labelToFind = unitLabel.toLowerCase();
+    const unitFamilyToFind = unitFamily ? unitFamily.toLowerCase():undefined;
+    const unitSystemToFind = unitSystem ? unitSystem.toLowerCase():undefined;
+
     for (const entry of UNIT_DATA) {
       if (unitFamily) {
-        if (entry.unitFamily !== unitFamily)
+        if (entry.unitFamily.toLowerCase() !== unitFamilyToFind)
```
Comment (calebmshafer) [inline core/frontend/src/BasicUnitsProvider.ts:27]: nit: Can you put these two if statements on the same line instead of breaking it up?

File: core/quantity/src/Parser.ts [hunk 3]
```diff
@@ -325,30 +352,51 @@ export class Parser {
     // common case where single value and single label are supplied
     if (tokens.length === 2) {
       if (tokens[0].isNumber && tokens[1].isString) {
-        const unit = await unitsProvider.findUnit(tokens[1].value as string, defaultUnit ? defaultUnit.unitFamily : undefined);
-        return new Quantity(unit, tokens[0].value as number);
+        const unit = await this.lookupUnitByLabel(tokens[1].value as string, format, unitsProvider);
```
Comment (calebmshafer) [inline core/quantity/src/Parser.ts:355]: There is a try/catch around the `lookupUnitByLabel` above, would the same workflow be appropriate here?  I'm reading through this and I don't see any handling for if the unit isn't found based on the token provided. Is it just expecting that it will throw in that case?

File: core/quantity/src/Parser.ts [hunk 4]
```diff
@@ -400,17 +472,17 @@ export class Parser {
     if (tokens.length === 1) {
       if (tokens[0].isNumber) {
```
Comment (calebmshafer) [inline core/quantity/src/Parser.ts:473]: nit: mind cleaning up the consecutive if statements here, makes it hard to parse through the code on review. Too many braces :)

File: core/quantity/test/Parsing.test.ts [hunk 5]
```diff
@@ -172,30 +171,64 @@ describe("Parsing tests:", () => {
...
+      { value: "12,345.345 FT", quantity: { magnitude: 12345.345, unitName: "Units.FT" } },
       { value: "3.6252e3 Miles", quantity: { magnitude: 3625.2, unitName: "Units.MILE" } },
       { value: "3.6252e-3 Miles", quantity: { magnitude: 0.0036252, unitName: "Units.MILE" } },
       { value: "3.6252e+3 Miles", quantity: { magnitude: 3625.2, unitName: "Units.MILE" } },
       { value: "-1 1/2FT", quantity: { magnitude: -1.5, unitName: "Units.FT" } },
       { value: "-2FT 6IN", quantity: { magnitude: -2.5, unitName: "Units.FT" } },
       { value: "1 1/2 FT", quantity: { magnitude: 1.5, unitName: "Units.FT" } },
       { value: "2FT 6IN", quantity: { magnitude: 2.5, unitName: "Units.FT" } },
-      { value: `2'-6"`, quantity: { magnitude: 2.5, unitName: "Units.FT" } },
+  /*    { value: `2'-6"`, quantity: { magnitude: 2.5, unitName: "Units.FT" } }, ambiguous case since ' can be Unit.FT or Units.ARC_MINUTE would require a Format unit to be defined. */
```
Comment (calebmshafer) [inline core/quantity/test/Parsing.test.ts:215]: Should this be a case where a preferred Unit is used?  I didn't quite follow on the initial review where the preferred unit comes from.

Comment (bsteinbk) [reply core/quantity/test/Parsing.test.ts:215]: No, not in this case because there is no unit defined in the Format... that means there is no unit to preference. In the test above this one the Format was a composite with unit[0] set to Units.FT so it had a preference.

Comment (calebmshafer) [reply core/quantity/test/Parsing.test.ts:215]: Ah okay, got it.

Comment (bsteinbk) [reply core/quantity/test/Parsing.test.ts:215]: I should also note that in imodeljs frontend and ui packages we use the other method of parsing where we pass in the persistence unit and in that case the persistence unit would serve as a fallback unit if the Format did not define one.

File: core/quantity/src/Parser.ts (hunk 3 above)
Comment (bsteinbk) [reply core/quantity/src/Parser.ts:355]: The try/catch must have been from an early implementation... there is nothing being thrown or expected to be thrown. I will remove the old try/catch.

File: core/quantity/src/Parser.ts (hunk 4 above)
Comment (bsteinbk) [reply core/quantity/src/Parser.ts:473]: They can't be combined in this case, because each if has separate else cases.

Comment (calebmshafer) [reply core/quantity/src/Parser.ts:473]: ah okay, missed that.

File: core/frontend/src/BasicUnitsProvider.ts (hunk 2 above)
Comment (bsteinbk) [reply core/frontend/src/BasicUnitsProvider.ts]: fixed.

File: core/quantity/src/Interfaces.ts (hunk 1 above)
Comment (bsteinbk) [reply core/quantity/src/Interfaces.ts:42]: Added note to breaking change section.

File: core/quantity/src/Parser.ts (hunk 3 above)
Comment (bsteinbk) [reply core/quantity/src/Parser.ts:352]: removed unneeded try/catch.

**Review by calebmshafer — APPROVED** [review]
